tests/ruby/tests.rb

```diff
@@ -201,3 +201,66 @@ def test_reload_pool_recycling
 end
 
 test_reload_pool_recycling
+
+
+
+def with_captured_stdout_stderr
+  sout = STDOUT.clone
+  serr = STDERR.clone
+  STDOUT.reopen("/tmp/out.txt", "w+")
+  STDERR.reopen("/tmp/err.txt", "w+")
+  yield
+  return File.read('/tmp/out.txt'), File.read('/tmp/err.txt')
+ensure
+  STDOUT.reopen(sout)
+  STDERR.reopen(serr)
+end
+
+def test_extended_protocol_pooler_errors
+  admin_conn = PG::connect("postgres://admin_user:admin_pass@127.0.0.1:6432/pgcat")
+
+  conf_editor = ConfigEditor.new
+  new_configs = conf_editor.original_configs
+
+  # shorter timeouts
+  new_configs["general"]["connect_timeout"] = 50
+  new_configs["general"]["ban_time"] = 1
+  new_configs["pools"]["sharded_db"]["users"]["0"]["pool_size"] = 1
+  new_configs["pools"]["sharded_db"]["users"]["1"]["pool_size"] = 1
+
+  conf_editor.with_modified_configs(new_configs) { admin_conn.async_exec("RELOAD") }
+
+  conn_str = "postgres://sharding_user:sharding_user@127.0.0.1:6432/sharded_db"
+  conn_under_test = PG::connect(conn_str)
+  50.times do
+    Thread.new do
```

Comment:
Curious why you need threads here, I thought `async_exec` is asynchronous so you could issue multiple queries without waiting for the result? Maybe it's not named correctly.

Reply:
My understanding is that `async_exec` blocks the caller but releases the GVL whereas `sync_exec` blocks the client and does not release the GVL. I haven't studied the code closely so I could be wrong.